bundles/org.openhab.binding.sunsynk/src/main/java/org/openhab/binding/sunsynk/internal/handler/SunSynkAccountHandler.java

```diff
@@ -0,0 +1,228 @@
+        updateConfiguration(configuration);
+        logger.debug("Account configuration updated : {}", configuration);
+    }
+
+    public void refreshAccount() {
+        Long expires_in;
+        Long issued_at;
+        Configuration configuration = editConfiguration();
+        try {
+            expires_in = ((Long) configuration.get("expires_in"));
+            issued_at = ((Long) configuration.get("issued_at"));
+        } catch (Exception e) {
+            expires_in = Long.parseLong(configuration.get("expires_in").toString());
+            issued_at = Long.parseLong(configuration.get("issued_at").toString());
+        }
+
+        if ((issued_at + expires_in) - Instant.now().getEpochSecond() > 30) { // 30 seconds
+            logger.debug("Account configuration token not expired.");
+            // logger.debug("Account fialed to refresh, token not expired. Trying re-auth");
+            // configAccount();
+            return;
+        }
+        logger.debug("Account configuration token expired : {}", configuration);
+        SunSynkAccountConfig accountConfig = getConfigAs(SunSynkAccountConfig.class);
+        String refreshtToken = configuration.get("refresh_token").toString();
+        Client sunAccount = refresh(accountConfig.getEmail(), refreshtToken);
+        Optional<APIdata> checkAPI = sunAccount.safeAPIData();
+        if (!checkAPI.isPresent()) { // API Data failed
+            logger.debug("Account fialed to refresh, likely a certificate path or maybe a password problem.");
+            updateStatus(ThingStatus.OFFLINE);
+            return;
+        }
+        APIdata apiData = checkAPI.get();
+        String newToken = apiData.getAccessToken();
+        APIdata.static_access_token = newToken;
+        configuration.put("access_token", newToken);
+        configuration.put("refresh_token", sunAccount.getRefreshTokenString());
+        configuration.put("expires_in", sunAccount.getExpiresIn());
+        configuration.put("issued_at", sunAccount.getIssuedAt());
+        updateConfiguration(configuration);
+        logger.debug("Account configuration refreshed : {}", configuration);
+    }
+
+    private Client authenticate(String username, String password) {
+        Gson gson = new Gson();
+        String response = "";
+        String httpsURL = makeLoginURL("oauth/token");
+        String payload = makeLoginBody(username, password);
+        Properties headers = new Properties();
+        try {
+            headers.setProperty("Accept", "application/json");
+            headers.setProperty("Content-Type", "application/json"); // may not need this.
+            InputStream stream = new ByteArrayInputStream(payload.getBytes(StandardCharsets.UTF_8));
+            response = HttpUtil.executeUrl("POST", httpsURL, headers, stream, "application/json", 2000);
+            Client API_Token = gson.fromJson(response, Client.class);
+            return API_Token;
+        } catch (IOException e) {
```

Comment:
There are more exceptions to catch. check other httprequests too.
Also add a timeout to all http requests

Reply:
Timeout  already  set = 2000
Improved exception handling.